Coaxing quality from genai blog post (#1936)
scala/scala-lang: blog/_posts/2026-06-29-quality-from-genai.md

---
layout: blog-detail
post-type: blog

> by: Bill Venners, Artima, Inc.

Reviewer: ```suggestion
by: Bill Venners (Artima, Inc.)
```

Another way we tried to raise the quality of the AI output prior to human review is to use _harnesses_ to orchestrate the AI work. First, we wrote a `todo-writer` application that reformatted doc comments for consistency and inserted text markers in the standard library source to precisely indicate where we wanted an AI to insert focused documentation. Then we created a script to drive the filling-in process. Finally, we attempted to keep the review tasks for humans to a reasonable size. The rest of this post describes these techniques in more detail.

## Technique 1: Push work onto deterministic code

> The [first pull request](https://github.com/scala/scala3/pull/24754) we submitted, which made the style of existing documentation more consistent, was done simply by asking an AI model to find the and make the changes, and reviewing the results. This prompting technique worked fine given the scope of those changes.

Reviewer: ```suggestion
The [first pull request](https://github.com/scala/scala3/pull/24754) we submitted, which made the style of existing documentation more consistent, was done simply by asking an AI model to find and make the changes, and then reviewing the results. This prompting technique worked fine given the limited scope of those changes.
```

We next ran `todo-writer` across the standard library to insert `TODO FILL IN` markers wherever existing documentation was missing any `@param`, `@tparam`, or `@return` tags. Here's an example:

```scala
/** Creates a new map obtained by updating this map with a given key/value pair.
 *  @tparam V1 TODO FILL IN
 *  @param key TODO FILL IN
 *  @param value TODO FILL IN
 *  @return TODO FILL IN
 */
def updated[V1 >: V](key: K, value: V1): CC[K, V1]
```

Each `TODO FILL IN` is one precise marker, which turned the AIs' job from the open-ended "find and improve the documentation" into the mechanical "replace this text with the right content." We could now focus the models on exactly those spots, with nothing about where to write or what needed writing left to their judgment.

> The principle underneath this technique is to use determinism wherever you can. Lean on a deterministic program for every part of the job that does not genuinely require judgment, and strip out any nondeterminism that is not essential. Every piece of the job you hand to deterministic code is a piece the AI cannot get wrong.

Reviewer: ```suggestion
The principle underneath this technique is to use determinism wherever you can. Lean on a deterministic program for every part of the job that does not genuinely require judgment, and strip out any nondeterminism that is not essential. _Every piece of the job you hand to deterministic code is a piece the AI cannot get wrong._
```

(This is such an important point that you might consider emphasizing it more strongly. Whether with italics, or bold, or giving it its own paragraph, or whatever you think works.)

## Technique 2: Enforce a deterministic process

> Another way we used determinism was by defining and enforcing a _process_ through a shell script that called into the AI model, rather than asking an AI to run the entire task. This kind of orchestration is often called an "AI harness," which both implies that an AI workhorse is being constrained as well as that it is being put to productive use.

Reviewer: ```suggestion
Another way we used determinism was by defining and enforcing a _process_ through a shell script that called into the AI model, rather than asking an AI to run the entire task. This kind of orchestration is often called an "AI harness," which both implies that an AI workhorse is being constrained as it is put to productive use.
```

> The cherry-pick was playing a critical role: it was the loop's iterator. Each source commit was exactly one unit of work, so cherry-picking popped the next file off the queue in order. The AI model never had to decide for itself which files needed attention or in what sequence, because that was settled when the queue was built. It also guaranteed that the output kept the same clean shape as the input, one commit per file, which later let us regroup the finished commits into review-sized pull requests. And because a file that had already been processed no longer appeared in the not-yet-applied range, the run was naturally restartable and could proceed in batches.

Reviewer: ```suggestion
The cherry-picking played a critical role: it was the loop's iterator. Each source commit was exactly one unit of work, so cherry-picking popped the next file off the queue in order. The AI model never had to decide for itself which files needed attention or in what sequence, because that was settled when the queue was built. It also guaranteed that the output kept the same clean shape as the input, one commit per file, which later let us regroup the finished commits into review-sized pull requests. And because a file that had already been processed no longer appeared in the not-yet-applied range, the run was naturally restartable and could proceed in batches.
```

> The benefit of this harness was to keep the AIs focused on small, isolated tasks inside a fixed create, review, and refine process. That process takes time, but not my time: I ran it overnight while I slept, with deliberate pauses between files to spread out my token usage. If you need a task to finish faster in wall-clock terms, you can design processes that run in parallel.

Reviewer: ```suggestion
The benefit of this harness was to keep the AIs focused on small, isolated tasks inside a fixed process of create, review, and refine. That process takes time, but not my time: I ran it overnight while I slept, with deliberate pauses between files to spread out my token usage. (Note that if you need a task to finish faster in wall-clock terms, you can design processes that run in parallel.)
```

## Technique 3: Give the AI a fresh, focused context

> Our script gave the AI a fresh context for every step, not just for every file. For a single file the writer filled the TODOs in a fresh context, the reviewer judged the result in a second fresh context, and the refiner applied the review in a third, with nothing carried between them except the file itself and the reviewer's written feedback. I kept each of those contexts as small as possible for two reasons: as context length increases, model quality often declines; and a larger context consumes more tokens on every turn. Although I was on a monthly subscription, I still did not want to burn more tokens than necessary.

Reviewer: is there a better phrase than "model quality"? "results quality"? it's not the model itself that's declining in quality

Author: You're right. It isn't model quality. I'll fix that...

Author: Improved it.

## What's next

Everything described here filled in tags on declarations that already had a Scaladoc comment. The larger job still ahead is to document the declarations that have none at all, and there are many: 5,894 program entities in the standard library currently have no Scaladoc comment whatsoever. That is a harder problem than supplying a missing `@param` or `@return`, because there is no existing summary to build on, so reaching high quality will take more than the single create, review, and refine pass we relied on here.

For that next batch I want to elaborate the review side of the loop. Instead of a single reviewer, I plan to run several reviews of each change, including reviews by different AI models, so that more eyes, and more kinds of eyes, weigh in before any human does. The aim is the same as before: push the quality as high as the machines can take it, so that by the time a human takes over, the documentation is already good.

> 

Reviewer: There's some standard fundraising boilerplate that all the Sovereign blog posts are supposed to end with; please borrow it from a previous post. (No opinion on whether it goes before or after "About the author".)

> ## About the Author

Reviewer: ```suggestion
## About the author
```

Author: Fixed.

Quality writing requires iteration. You write a draft, review it (or have someone else review it), then improve it. You repeat this process until you're happy with the quality. One way to achieve quality with generative AI, therefore, is to delegate the writing part to the AI model and review it yourself, but this doesn't scale. An AI model can generate content at a much greater rate than you can review it. Because of this, we decided to use the ["LLM as a judge"](https://en.wikipedia.org/wiki/LLM-as-a-Judge) technique in an attempt to raise the quality of the output as high as possible prior to asking humans to review it.

> Another way we tried to raise the quality of the AI output prior to human review is to use _harnesses_ to orchestrate the AI work. First, we wrote a `todo-writer` application that reformatted doc comments for consistency and inserted text markers in the standard library source to precisely indicate where we wanted an AI to insert focused documentation. Then we created a script to drive the filling-in process. Finally, we attempted to keep the review tasks for humans to a reasonable size. The rest of this post describes five techniques that formed our overall approach.

Reviewer: ```suggestion
Another way we tried to raise the quality of the AI output prior to human review is to use _harnesses_ to orchestrate the AI work. First, we wrote a `todo-writer` application that reformatted doc comments for consistency and inserted text markers in the standard library source to precisely indicate where we wanted an AI to insert focused documentation. Then we created a script to drive the filling-in process. Finally, we attempted to keep the review tasks for humans to a reasonable size.

The rest of this post describes five techniques that formed our overall approach.
```
